### PR TITLE
Fix double counting V3 groups bug

### LINKED ISSUE #1228: Double counting groups in V3
### Zarr version

2.13.3

### Numcodecs version

n/a

### Python Version

3.10

### Operating System

Mac

### Installation

poetry

### Description

I have discovered that certain zarr methods double count groups in V3. The methods `Group.group_keys()` and `Group.groups()` return the group twice when the group contains children.

Instead, each group should only be returned once.

### Steps to reproduce

The following test should pass

```python
import zarr

def test_double_counting_group_v3():
    store = zarr.MemoryStoreV3()
    root_group = zarr.group(store=store, zarr_version=3)
    sub_group = root_group.create_group("foo")
    array = sub_group.create("bar", shape=10, dtype="i4")
    # this works and is in spec (although the spec says it should be "foo/")
    assert store.listdir("meta/root") == ['foo', 'foo.group.json']
    # these should work but don't: instead we get foo twice
    assert list(root_group.group_keys()) == ['foo']
    assert list(root_group.groups()) == [("foo", sub_group)]
```

What is happening is that the the `listdir` call here is returning both `foo` and `foo.group.json`. Because `contains_group('foo')` is valid, the group is returned twice.

https://github.com/zarr-developers/zarr-python/blob/ce129a560d48854aee533bb2699a3f28b396bc22/zarr/hierarchy.py#L513-L521

If others agree this is a bug, I will submit a PR to fix it.

### Additional output

_No response_

### PR DESCRIPTION
**Pull request by rabernat**
Fixes #1228

TODO:
* [x] Add unit tests and/or doctests in docstrings
* ~Add docstrings and API docs for any new/modified user-facing classes and functions~
* ~New/modified features documented in docs/tutorial.rst~
* [x] Changes documented in docs/release.rst
* [x] GitHub Actions have all passed
* [x] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by rabernat** [comment]
Is anyone up for reviewing this simple PR?

File: docs/release.rst [hunk 1]
```diff
@@ -14,6 +14,12 @@ Unreleased
     # .. warning::
     #    Pre-release! Use :command:`pip install --pre zarr` to evaluate this release.
 
+* Fix bug that caused double counting of groups in ``groups()`` and ``group_keys()``
+  methods with V3 stores.
+  By :user:`Ryan Abernathey <rabernat>` :issue:`1228`.
+
+.. _release_2.13.2:
```
Comment (joshmoore) [inline docs/release.rst:21]: This might end up being redundant:

https://github.com/zarr-developers/zarr-python/blob/main/docs/release.rst#2132

**Review by joshmoore — APPROVED** [review]
Barring the minor release.rst issue (which will need work anyway for the release), this makes sense to me.

File: docs/release.rst (hunk 1 above)
Comment (rabernat) [reply docs/release.rst:21]: I can't tell if you are requesting changes to my PR or not. Could you be more explicit?

Comment (joshmoore) [reply docs/release.rst:21]: Sorry. I updated the branch and there doesn't look to be a conflict. (My expectation was there were going to be two)

**Pull request merged**